src/frontend/src/binder/insert.rs

```diff
@@ -135,151 +130,234 @@ impl Binder {
+                        //      insert into t values (7)
+                        // this kind of usage is fine, null values will be provided implicitly.
+                        std::cmp::Ordering::Greater => None,
+                        // e.g. create table t (a int, b real)
+                        //      insert into t values (7, 13, 17)
+                        std::cmp::Ordering::Less => {
+                            Some("INSERT has more expressions than target columns")
+                        }
+                    };
 
-        if let Some(msg) = err_msg && !target_table_col_indices.is_empty() {
-            return Err(RwError::from(ErrorCode::BindError(
-                msg.to_string(),
-            )));
-        }
+                    if let Some(msg) = err_msg {
+                        return Err(RwError::from(ErrorCode::BindError(msg.to_string())));
+                    }
 
-        // Check if column was used multiple times in query e.g.
-        // insert into t1 (v1, v1) values (1, 5);
-        let mut uniq_cols = target_table_col_indices.clone();
-        uniq_cols.sort_unstable();
-        uniq_cols.dedup();
-        if target_table_col_indices.len() != uniq_cols.len() {
-            return Err(RwError::from(ErrorCode::BindError(
-                "Column specified more than once".to_string(),
-            )));
+                    let values = self.bind_values(values, Some(expected_types))?;
+                    let body = BoundSetExpr::Values(values.into());
+                    (
+                        BoundQuery {
+                            body,
+                            order: vec![],
+                            limit: None,
+                            offset: None,
+                            with_ties: false,
+                            extra_order_exprs: vec![],
+                        },
+                        vec![],
+                    )
+                }
+                query => {
+                    let bound = self.bind_query(query)?;
+                    let actual_types = bound.data_types();
+                    let cast_exprs = match expected_types == actual_types {
+                        true => vec![],
+                        false => Self::cast_on_insert(
+                            &expected_types,
+                            actual_types
+                                .into_iter()
+                                .enumerate()
+                                .map(|(i, t)| InputRef::new(i, t).into())
+                                .collect(),
+                        )?,
+                    };
+                    (bound, cast_exprs)
+                }
+            };
+            let insert = BoundInsert {
+                table_id,
+                table_version_id,
+                table_name,
+                owner,
+                row_id_index,
+                column_indices: (0..columns_to_insert.len()).collect::<Vec<usize>>(),
```

Comment:
when user doesn't specify target column identifiers, the column indices should be ordinal

Reply:
should I edit this into the code comments?